### PR TITLE
Added support for Yada (rx/yada algorithm)

### PR DESCRIPTION
**Pull request by pdxwebdev**
Hello, YadaCoin would like to be added to the master branch of xmrig.

### CODE REVIEW

**Comment by xmrig** [comment]
This is some kind of testnet? network hashrate is below 10 kH/s.
Thank you.

**Comment by pdxwebdev** [comment]
> This is some kind of testnet? network hashrate is below 10 kH/s.
> Thank you.

Looking at https://miningpoolstats.stream/yadacoin, I would agree the hash rate reported should be higher for the number of miners. We'll look into that. Thanks.

However, for this PR, is there any additional information you need from us? 

Test command:
`./xmrig --donate-level 5 -o pool.yadacoin.io:3333 -u 1ArsFNcc5fU3cfSUiNJCu6LhT8CeZgtEcC -a rx/yada`

Getting about 2.5 kH/s

**Comment by pdxwebdev** [comment]
We fixed our hashrate reporting issue. Our network hashrate is now properly reported.

**Comment by ajstead** [comment]
I can add my vote for this - I'm currently mining Yada, network hash rate is good, and listed on SafeTrade/

**Comment by tybiboune** [comment]
Yadacoin is nothing like a "testnet"... the hashrate is fine and it's one of the VERY RARE projects with a REAL usecase that solves a REAL problem using a very SMART way. Definitely deserves being merged.

**Comment by username1565** [comment]
Wow, rx/yada in GPU's version of xmrig? I saw this only in XmrigCC, but it's for CPU: https://github.com/Bendr0id/xmrigCC/pull/343
and in SRBMiner-Multi: https://github.com/doktor83/SRBMiner-Multi/blob/9ce7a25bbe6883d7f660de6e012c9d4077a3088d/Readme#L44
but it's for AMD.

I see this pull-request have a some unresolved conflicts,
it would be nice to fix this, to make Nvidia GPU-miner with algo `rx/yada` (`randomYADA`).

**Comment by MrSMiTHex** [comment]
> This is some kind of testnet? network hashrate is below 10 kH/s. Thank you.

Yadacoin uses rx/yada for about two years. Is supported by XmrigCC and SRB. What is needed to add its support to Xmrig?

**Comment by pdxwebdev** [comment]
@xmrig Very simple merge. Already being used in another fork of xmrig. Think we can move forward?

**Comment by pdxwebdev** [comment]
Screenshot from 2021-11-25 00-55-04

**Comment by username1565** [comment]
Wow! I see conflicts was been resolved, in this pull-request.
@pdxwebdev , is this working on CPU only, or with GPU's too?
I see OpenCL, and CUDA is disabled, on your screen, and only CPU sending the resolved shares.

**Comment by pdxwebdev** [comment]
@username1565 I have implemented everything required for CUDA and OpenCL. I'll provide a screenshot when I'm able to test. If someone else is able to test either of those in the meantime, I would appreciate it.

**Comment by Spudz76** [comment]
@pdxwebdev Please rebase on `dev`; Also run `node ./scripts/generate_cl.js` and commit the resulting regenerated header files.

**Comment by Spudz76** [comment]
@pdxwebdev Consult [my branch `pdx-addRXYADA`](https://github.com/Spudz76/xmrig/commit/f8c26b46aec1c34ed1d27065c51c5652c69641de) where I have already fixed everything that collided with `dev` (and improved / arranged some other bits).

**Comment by Spudz76** [comment]
@pdxwebdev Also check over in my [xmrig-cuda/pdx-addRXYADA](https://github.com/Spudz76/xmrig-cuda/tree/pdx-addRXYADA) where I wired up maybe-working other half of the CUDA support

**Comment by pdxwebdev** [comment]
@Spudz76  Awesome! Thanks for doing that.

It may not be getting an accepted share because the RandomX Yada configuration modifications for CUDA / OpenCL. 

Do these variables need to be set elsewhere?
ArgonSalt = "RandomXYadaCoin\x03";
SuperscalarLatency = 150;
ArgonIterations = 4;

**Comment by Spudz76** [comment]
I set those in the CUDA plugin side, should be valid if I happened to find a share above the diff.

I didn't have CPU enabled the combination of the two should find something but doesn't exactly prove out the CUDA (accepts would likely be just from cpu)

If there were any pools with selectable diff or autodiff it would help.  GPUs just won't crank much hash with rx/anything.

Testing the OpenCL next...

**Comment by Spudz76** [comment]
OpenCL seems to work; although same as CUDA the diff is too high to catch a validation accept (or reject).  Diff would need to be ~9150 to see something.  69905 is good for ~2330H/s

```
 * POOL #1      pool.yadacoin.io:3333 algo rx/yada
 * COMMANDS     hashrate, pause, resume, results, connection
 * HTTP API     0.0.0.0:10084 
 * ADL          press e for health report
 * OPENCL       #0 AMD Accelerated Parallel Processing/OpenCL 2.1 AMD-APP (3143.9)
 * OPENCL GPU   #0 01:00.0 AMD Radeon (TM) RX 480 Graphics (Ellesmere) 1303 MHz cu:36 mem:4088/4088 MB
[2021-11-25 20:51:08.361]  net      use pool pool.yadacoin.io:3333  54.172.163.172
[2021-11-25 20:51:08.361]  net      new job from pool.yadacoin.io:3333 diff 69905 algo rx/yada height 247235
[2021-11-25 20:51:08.361]  cpu      use argon2 implementation SSSE3
[2021-11-25 20:51:08.361]  randomx  init dataset algo rx/yada (1 threads) seed 4181a493b397a733...
[2021-11-25 20:51:08.362]  randomx  allocated 2336 MB (2080+256) huge pages 0% 0/1168 +JIT (0 ms)
[2021-11-25 20:51:47.567]  randomx  dataset ready (39205 ms)
[2021-11-25 20:51:47.567]  opencl   use profile  rx  (2 threads) scratchpad 2048 KB
|  # | GPU |  BUS ID | INTENSITY | WSIZE | MEMORY | NAME
|  0 |   0 | 01:00.0 |       448 |     8 |    896 | AMD Radeon (TM) RX 480 Graphics (Ellesmere)
|  1 |   0 | 01:00.0 |       448 |     8 |    896 | AMD Radeon (TM) RX 480 Graphics (Ellesmere)
[2021-11-25 20:51:47.789]  opencl   GPU #0 compiling...
[2021-11-25 20:52:04.339]  net      use pool pool.yadacoin.io:3333  54.172.163.172
[2021-11-25 20:52:04.339]  net      new job from pool.yadacoin.io:3333 diff 69905 algo rx/yada height 247235
[2021-11-25 20:52:04.717]  net      use pool pool.yadacoin.io:3333  54.172.163.172
[2021-11-25 20:52:04.717]  net      new job from pool.yadacoin.io:3333 diff 69905 algo rx/yada height 247236
[2021-11-25 20:52:11.100]  opencl   GPU #0 compilation completed (23311 ms)
[2021-11-25 20:52:11.122]  opencl   READY threads 2/2 (23554 ms)
[2021-11-25 20:52:51.824]  opencl   #0 01:00.0  41W 35C    0RPM 769/1750MHz
[2021-11-25 20:52:53.400]  miner    speed 10s/60s/15m 324.4 88.63 n/a H/s max 343.9 H/s
[2021-11-25 20:53:31.369]  net      use pool pool.yadacoin.io:3333  54.172.163.172
[2021-11-25 20:53:31.369]  net      new job from pool.yadacoin.io:3333 diff 69905 algo rx/yada height 247236
[2021-11-25 20:53:31.886]  net      use pool pool.yadacoin.io:3333  54.172.163.172
[2021-11-25 20:53:31.886]  net      new job from pool.yadacoin.io:3333 diff 69905 algo rx/yada height 247237
[2021-11-25 20:53:57.557]  opencl   #0 01:00.0  41W 40C    0RPM 751/1750MHz
[2021-11-25 20:53:59.072]  miner    speed 10s/60s/15m 325.4 320.0 n/a H/s max 348.2 H/s
```

**Comment by pdxwebdev** [comment]
@Spudz76  Let's make the following modification.
`Job.cpp, line 116`
`- m_target = 0x0000F00000000000ULL + strtoull(target, nullptr, 16);`
`+ m_target = strtoull(target, nullptr, 16);`

And then use this pool: 18.214.218.185:3333

**Comment by Spudz76** [comment]
Yes that pool is nice diff, now I can see CUDA doesn't actually work yet (compute error always, when a result should occur)

**Comment by Spudz76** [comment]
Same deal with OpenCL.  hmmmm.  CPU does work, still.  Tried all with or without that m_target patch.

**Comment by pdxwebdev** [comment]
@Spudz76 I can get it working if I drop the pool diff to 1.
Screenshot from 2021-11-26 17-22-15

What appears to be happening is:

CheckHash is immediately called after a hash is calculated
https://github.com/Spudz76/xmrig/blob/master/src/net/JobResults.cpp#L138

CheckHash compares the hash to the target. It sees that the hash is above the target so it logs a compute error.
https://github.com/Spudz76/xmrig/blob/master/src/net/JobResults.cpp#L105

Shouldn't it simply throw out the hash and keep trying to find a hash that _is_ less than the target instead of logging an error?

**Comment by Spudz76** [comment]
Interesting, and what I suspected.

The GPU backend thinks it found a result above target or it wouldn't have sent it up to be rechecked.  But it wasn't actually above target.  Or it was and the verification is incorrect... I get almost no compute errors when the diff is gigantic, because no results are found.  I think the results are valid but the verification is wrong.

**Comment by pdxwebdev** [comment]
@siudz76 I agree, and I also checked the hashes being received by the pool. They are valid.

**Comment by pdxwebdev** [comment]
@xmrig, @Spudz76 changes are now merged into this branch (thank you @Spudz76). I also executed `node ./scripts/generate_cl.js` and committed the resulting headers.

Is there anything else that needs to be done? Thanks!

**Comment by username1565** [comment]
@pdxwebdev , @xmrig , @ajstead , @tybiboune , @MrSMiTHex, @Spudz76 
Hello, guys!
Somebody know how to setup `rx/yada` xmrig on HiveOS?
I just go into my account on https://the.hiveos.farm and try to creat `Flight Sheet` for YadaCoi (YDA),
then select xmrig, but I do not see this algo, there: 
image

Maybe need to add this version of miner, as custom-miner?
like in this guide: https://hiveos.farm/getting_started-start_custom_miner/
I did try to use `.tar.gz` archive here from latest release: https://github.com/xmrig/xmrig/releases/tag/v6.16.2
this one https://github.com/xmrig/xmrig/archive/refs/tags/v6.16.2.tar.gz
but this is not working, and this ask for some files, like this:
https://github.com/MoneroOcean/hiveos/tree/main/mo_xmrig

Maybe there is possible to create another branch with that HiveOS config-files, or something else.

**Comment by username1565** [comment]
Is there possible, according `donate-level`, to pay donations in YadaCoins, and do mining that donations, using algo `rx/yada`, just to do not let auto-switch the miner on `another algo` (when donate mining starting), and do not switch that algo, while overclock-settings applied for `rx/yada`? I saw earlier GPU driver-crashes, when donate mining starting with another algo, from target algo, when GPU is overclocked for this target algo.

**Comment by Aurok85** [comment]
Please add support for Yada to xmrig! I love xmrig and want to use your miner for every coin I choose to mine, it is THE miner for CPU mining.. period!

**Comment by pdxwebdev** [comment]
Hello @Spudz76 , @xmrig,

We did some great work to improve this PR before the holidays and I continue to keep this PR up-to-date with the dev branch.

What is still left outstanding for merging this PR?

**Comment by pdxwebdev** [comment]
Checking in again @xmrig.

At the time of writing this:
- Our network hash rate is at around 1.2 MH/s
- 130 miners
- We just got featured by rabid mining: https://www.youtube.com/watch?v=T6f7CdzbTMI

It would be fantastic if we could get added to this repository or at least know the minimum metrics we need to hit.

Thanks, Matt

**Comment by pdxwebdev** [comment]
Hello @Spudz76, @xmrig 

I see you are referring most projects to implement their algo on 3rd party forks. This makes things for smaller project quite difficult to make their project minable on the various mining platforms that only track the xmrig main branch.

Without inclusion into the xmrig main branch, it is very limiting for our project.

Is there something we can do to be included in the xmrig main repo?

Thanks

**Comment by Spudz76** [comment]
I approve the code obviously, it's been in my own dev tree for a long time now.

**Comment by pdxwebdev** [comment]
> I approve the code obviously, it's been in my own dev tree for a long time now.

Thank you @Spudz76 . I'm wondering who / what else is required for merge?

**Comment by Spudz76** [comment]
Usually, @SChernykh comes through and refuses to merge at least.

**Comment by pdxwebdev** [comment]
@SChernykh I'm willing to make an XMR donation if we can merge this PR. Let me know what you think is fair.

**Comment by Rakni1988** [comment]
Can you add rx/yada to xmrig, many things changed from last request :)

**Comment by pdxwebdev** [comment]
Hello @xmrig @SChernykh @Spudz76 

It sure would be swell of y'all to add our lil' ol' project to this here repo. 🤠

**Comment by pdxwebdev** [comment]
In all seriousness @xmrig @SChernykh @Spudz76

We have one patent issued and one pending for our crypto innovations. Not being added to this main repo for xmrig is hindering our ability to reach a broader audience of miners on a variety of mining platforms.

If there is anything at all I can do to move this forward, I will do it and I'll have the support of several others from our community ready and willing to put in any amount of effort.

Please give us your sincere consideration.

**Comment by TomPelsz** [comment]
@xmrig @SChernykh @Spudz76 
Yada coin community would like to have it on main branch of xmrig! This blockchain is gaining more and more traction and @pdxwebdev is super dedicated dev!

**Comment by pdxwebdev** [comment]
Why was KevaCoin added and not YadaCoin? Their hashrate is a 1/10 of ours.
https://miningpoolstats.stream/kevacoin

**Comment by TomPelsz** [comment]
Come on guys, Yada coin is growing. It has now legit business using this blockchain, tons of development, steady hashrate and volume on CEX's is picking up. It deserves merging into randomX.

**Comment by Spudz76** [comment]
@TomPelsz I have less than zero to do with anything here other than contributing, and I'm already subscribed to this thread.

**Comment by SChernykh** [comment]
> Why was KevaCoin added and not YadaCoin? Their hashrate is a 1/10 of ours.

https://github.com/xmrig/xmrig/pull/1578

It was added in March 2020 when RandomX coins were rare, and Keva was much bigger coin back then. We should remove it now.

**Comment by pdxwebdev** [comment]
@SChernykh,

RandomX coins are still rare, especially now that the main RandomX variant has an ASIC miner capable of 212Kh/s.

I don't understand the issue with adding us. All the coding has been done and available for years. All you would need to do is review, click merge, and we could move forward.

- Saying you don't add any new coins while also recommending each coin modify the RX configuration is contradictory and hurts the respective projects. 
- Support for XMRig forks on mining and pool platforms is abysmal and significantly hinders project growth.

I understand that there are a lot of BS projects that come and go, but we aren't going anywhere. In fact, we are thriving. 

So please give us the go-ahead, and I will update this PR for your review and merge.

**Comment by SChernykh** [comment]
> RandomX variant has an ASIC miner capable of 212Kh/s.

It is not an ASIC. It's 18 RISC-V CPUs packed into one box, and their efficiency is on par with best CPU rigs (Ryzen 7900X/7950X and EPYC CPUs).

**Comment by pdxwebdev** [comment]
> It's 18 RISC-V CPUs packed into one box, and their efficiency is on par with best CPU rigs (Ryzen 7900X/7950X and EPYC CPUs).

To match the performance, you'd have to pay more than double.

The original intent of RandomX was to make mining accessible to smaller miners using just a regular CPU, not $2,000 rigs that can do 200Kh/s.

I don't understand the pushback here. This PR is not an aberration for this repo. Let's just get it done.

**Comment by SChernykh** [comment]
Antminers "ASICs" were sold waaaaay below their real price. SG2042 CPUs they have inside cost more than $1000 **each** if they sold them as a part of regular PC builds. Add 18 DDR4 memory sticks to that - they use server grade RAM there which is not cheap ($50 per stick, that's $900 only for memory).

TLDR Antminer "ASICs" were used to mine XMR for 2 years and then dumped on users for a max price they could pay, because the alternative for Bitmain was to throw them away (they were becoming unprofitable).

**Comment by SChernykh** [comment]
> I don't understand the pushback here. This PR is not an aberration for this repo. Let's just get it done.

The reason is simple, if a coin is too small (emission per day), but with a different algorithm, it will require extra resources each month (dev fee servers etc.), and dev fee from that emission will not cover it. So yes, while it's possible to add any coin under the sun, it will be a burden (net negative) to support.

**Comment by pdxwebdev** [comment]
> TLDR Antminer "ASICs" were used to mine XMR for 2 years and then dumped on users for a max price they could pay, because the alternative for Bitmain was to throw them away (they were becoming unprofitable).

This contradicts direct feedback from my mining community. They are actually telling me to dump RandomX and go with GR. Clearly, it is still a concern and influences their decision on where to allocate their hashrate.

However, that is immaterial. The point is that we have followed the documented instructions to modify the RX config for our project, and it would take little effort on your part to review and merge our code.

**Comment by pdxwebdev** [comment]
> The reason is simple, if a coin is too small (emission per day), but with a different algorithm, it will require extra resources each month (dev fee servers etc.), and dev fee from that emission will not cover it. So yes, while it's possible to add any coin under the sun, it will be a burden (net negative) to support.

Ok, so it's about money; that's fine. However, my main contention is that our emission would be substantially higher if we were added to this repo. We've had countless miners come and go because they would have to change 50+ rigs to accommodate a new XMRig fork.

If you want an upfront "listing fee," just name your price. I'm sure we could gather the funds to get this across the finish line.

**Comment by SChernykh** [comment]
There's no listing fee, we don't do that.

**Comment by pdxwebdev** [comment]
> There's no listing fee, we don't do that.

Ok, how about this.

Merge us and give us 6 months. In the meantime, we can supplement any fees related to the support of our coin should it fall short of your break-even point. If we cannot demonstrate any growth in that time period, simply remove us along with KevaCoin.

**Comment by xmrig** [comment]
@pdxwebdev We considered giving you a chance. Please update this PR or make a new one without changing `src/version.h` and `CHANGELOG.md`. Now, it conflicts with the dev branch.
Thank you.

**Comment by pdxwebdev** [comment]
@xmrig @SChernykh The changes are in, tested, are now ready for your review.

**Review by SChernykh — APPROVED** [review]

**Comment by xmrig** [comment]
@pdxwebdev The `pool.yadacoin.io` pool sends the target as `-000000000000000000000000000000000000483a53b8e4b9`, which is not expected, so the target parse failed.

**Comment by pdxwebdev** [comment]
> @pdxwebdev The `pool.yadacoin.io` pool sends the target as `-000000000000000000000000000000000000483a53b8e4b9`, which is not expected, so the target parse failed.

No problem, we just released support for xmrig 6. 

Please test pool.yadacoin.io again.

**Comment by Spudz76** [comment]
@pdxwebdev Any chance of poking TradeOgre again, to list this?

**Comment by pdxwebdev** [comment]
> @pdxwebdev Any chance of poking TradeOgre again, to list this?

I'll submit a listing request.

**Comment by SChernykh** [comment]
@pdxwebdev pool.yadacoin.io doesn't send `"algo":"rx/yada"` in new jobs, so XMRig disconnects (it can't parse `target` because it doesn't use rx/yada parsing code with this one):
> [2024-08-07 00:32:49.287] [pool.yadacoin.io:3333] received (825 bytes): "{"jsonrpc": "2.0", "method": "job", "params": {"blob": "353137323239383335353030333962366261386233313461343561653035633538313165643964323265366665626232333465363664303365623733393836343130313865313233393736313435303134383464653365613764373833353731633430663137663338376536303938303436386231303934666336653030646666336431386261386236613035303030303030000000008891143030303030303038373636666536633639336663383030303030303030303030303030303030303030303030303030303030303030303030303030303030303032613830316262663864383238636431393166393561333266316333383830633933393331303961653437393363353066396537616230386339666630623832", "job_id": "eb860938-a014-42a3-b21f-ba958e1ec15a", "target": "0xa7c5ac471b47", "seed_hash": "4181a493b397a733b083639334bc32b407915b9a82b7917ac361816f0a1f5d4d", "height": 501484, "extra_nonce": "889114"}}"

**Comment by pdxwebdev** [comment]
> @pdxwebdev pool.yadacoin.io doesn't send `"algo":"rx/yada"` in new jobs, so XMRig disconnects (it can't parse `target` because it doesn't use rx/yada parsing code with this one):
> 
> > [2024-08-07 00:32:49.287] [pool.yadacoin.io:3333] received (825 bytes): "{"jsonrpc": "2.0", "method": "job", "params": {"blob": "353137323239383335353030333962366261386233313461343561653035633538313165643964323265366665626232333465363664303365623733393836343130313865313233393736313435303134383464653365613764373833353731633430663137663338376536303938303436386231303934666336653030646666336431386261386236613035303030303030000000008891143030303030303038373636666536633639336663383030303030303030303030303030303030303030303030303030303030303030303030303030303030303032613830316262663864383238636431393166393561333266316333383830633933393331303961653437393363353066396537616230386339666630623832", "job_id": "eb860938-a014-42a3-b21f-ba958e1ec15a", "target": "0xa7c5ac471b47", "seed_hash": "4181a493b397a733b083639334bc32b407915b9a82b7917ac361816f0a1f5d4d", "height": 501484, "extra_nonce": "889114"}}"

Ok, I just released a fix.

"algo": "rx/yada" is now sent with new jobs.

**Comment by pdxwebdev** [comment]
@Spudz76 we just got listed on TradeOgre!

**Comment by xmrig** [comment]
@pdxwebdev https://github.com/xmrig/xmrig/releases/tag/v6.22.0
